glommio/src/net/tcp_socket.rs

```diff
@@ -0,0 +1,1022 @@
+                    if !self.source_rx.as_ref().unwrap().has_result() {
+                        self.source_rx
+                            .as_ref()
+                            .unwrap()
+                            .add_waiter(cx.waker().clone());
+                        return Poll::Pending;
+                    } else {
+                        self.rx_yolo = true;
+                    }
+                }
+                Some(source) => {
+                    self.rx_yolo = true;
+                    match source.extract_source_type() {
+                        SourceType::SockRecv(mut buf) => {
+                            let bytes = source.take_result().unwrap();
+                            match bytes {
+                                Err(x) => return Poll::Ready(Err(x)),
+                                Ok(sz) => {
+                                    let mut buf = buf.take();
+                                    buf.as_mut().unwrap().trim_to_size(sz);
+                                    self.rx_buf = buf;
+                                    return Poll::Ready(Ok(sz));
+                                }
+                            }
+                        }
+                        _ => unreachable!(),
+                    }
+                }
+            }
+        }
+    }
+
+    fn write_dma(&mut self, cx: &mut Context<'_>, buf: DmaBuffer) -> Poll<io::Result<usize>> {
+        let source = self.source_tx.take();
+        match source {
+            None => {
+                let source = self
+                    .reactor
+                    .new_source(self.stream.as_raw_fd(), SourceType::SockSend(buf));
+                self.reactor.sys.send(&source, iou::MsgFlags::empty());
+                if let Err(err) = self.reactor.sys.rush_dispatch(&source) {
+                    return Poll::Ready(Err(err));
+                }
+                match source.take_result() {
+                    None => {
+                        source.add_waiter(cx.waker().clone());
+                        self.source_tx = Some(source);
+                        Poll::Pending
+                    }
+                    Some(res) => {
+                        self.tx_yolo = true;
+                        Poll::Ready(res)
+                    }
+                }
+            }
+            Some(source) => {
+                self.tx_yolo = true;
+                let res = source.take_result().unwrap();
+                Poll::Ready(res)
+            }
+        }
+    }
+}
+
+impl AsyncBufRead for TcpStream {
+    fn poll_fill_buf<'a>(
+        mut self: Pin<&'a mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<io::Result<&'a [u8]>> {
+        loop {
+            if self.rx_buf.as_ref().is_some() {
+                let this = self.project();
+                return Poll::Ready(Ok(this.rx_buf.as_ref().unwrap().as_bytes()));
+            } else if let Err(err) = ready!(self.poll_replenish_buffer(cx, 8192)) {
```

Comment:
I was also thinking of making the buffer size configurable. Do you see any value in that?
It is not configurable for any of the other crates I see around that deal with networking.

Reply:
Yeah being able to tweak buffer sizes would interest me personally. But whether its needed now, compared to other stuff, that's another question entirely. Might be worth just sticking as an issue to be worked on later as a 'good first issue' type thing.

Reply:
easiest is to allow for buffer size picking, with a default on 8192.

I haven't pushed it yet because I am playing with buffer selection but will soon